apps/postgresql.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apps
+
+import (
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
+
+	"strings"
+)
+
+type MetricsReceiverPostgresql struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+
+	confgenerator.MetricsReceiverShared `yaml:",inline"`
+
+	Endpoint string `yaml:"endpoint" validate:"omitempty,hostname_port|startswith=/"`
+
+	Insecure           *bool  `yaml:"insecure" validate:"omitempty"`
+	InsecureSkipVerify *bool  `yaml:"insecure_skip_verify" validate:"omitempty"`
+	CertFile           string `yaml:"cert_file" validate:"omitempty"`
+	KeyFile            string `yaml:"key_file" validate:"omitempty"`
+	CAFile             string `yaml:"ca_file" validate:"omitempty"`
+
+	Password  string   `yaml:"password" validate:"omitempty"`
+	Username  string   `yaml:"username" validate:"omitempty"`
+	Databases []string `yaml:"databases" validate:"omitempty"`
+}
+
+// Actual socket is /var/run/postgresql/.s.PGSQL.5432 but the lib/pq go module used by
+// the underlying receiver expects it like this
+const defaultpostgresqlUnixEndpoint = "var/run/postgresql/:5432"
+
+func (r MetricsReceiverPostgresql) Type() string {
+	return "postgresql"
+}
+
+func (r MetricsReceiverPostgresql) Pipelines() []otel.Pipeline {
+	transport := "tcp"
+	if r.Endpoint == "" {
+		transport = "unix"
+		r.Endpoint = defaultpostgresqlUnixEndpoint
+	} else if strings.HasPrefix(r.Endpoint, "/") {
+		transport = "unix"
+		endpointParts := strings.Split(r.Endpoint, ".")
+		r.Endpoint = strings.TrimLeft(endpointParts[0], "/") + ":" + endpointParts[len(endpointParts)-1]
+	}
+
+	if r.Username == "" {
+		r.Username = "postgres"
+	}
+
+	cfg := map[string]interface{}{
+		"collection_interval": r.CollectionIntervalString(),
+		"endpoint":            r.Endpoint,
+		"username":            r.Username,
+		"password":            r.Password,
+		"transport":           transport,
+	}
+
+	if transport == "tcp" {
```

Comment:
Similar here, multiple application will have logic for `TLS`. Can we use a shared helper? Something similar to `confgenerator.MetricsReceiverShared`. So that future applications can use the same helpful. This helps boost consistency across applications. And it's overall easier to maintain in the long term.

Reply:
The only thing that might differ between applications is default TLS, so I'll try to factor that in.